blog/2017-09-25-streaming-to-another-database.adoc

```diff
@@ -0,0 +1,222 @@
+= Streaming data to a downstream database
+Jiri Pechanec
+:awestruct-tags: [ mysql, postgres, smt, example ]
+:awestruct-layout: blog-post
+
+Recently there was a short https://www.confluent.io/blog/simplest-useful-kafka-connect-data-pipeline-world-thereabouts-part-1/[series] of blogposts on https://www.confluent.io/[Confluent's] blog.
+
+In this blogpost we will try to create a simple scenario: we are going to have a MySQL database instance running as a master database that feeds a Kafka broker with events representing a source of truth.
+Next we will add PostgreSQL database that will listen one of the topics and will replicate the state of the MySQL instance to its own data.
+
+We will also demonstrate a new functionality coming with version link:2017/09/21/debezium-0-6-0-released[0.6.0]: a single message transform for link:/docs/configuration/event-flattening[CDC Event Flattening].
+
+== Topology
+The general topology for this scenario is displayed on the following picture:
+
+.A General topology
+[#img-general]
+[caption="Figure 1: "]
+image::dbz-to-jdbc.svg[Scenario topology]
+
+To simplify the setup a little bit we will use only one Kafka Connect instance that will contain all connectors so the single instance will serve as an event producer and an event consumer.
+
+.A Simplified topology
+[#img-general]
+[caption="Figure 2: "]
+image::dbz-to-jdbc-simplified.svg[Scenario topology]
+
+== Configuration
+We will use this https://github.com/debezium/debezium-examples/tree/master/unwrap-smt[compose] for a fast deployment of the demo.
+The deployment consists of following images
+
+* Apache ZooKeeper
+* Apache Kafka
+* An enriched Kafka Connect image with changes
+** PostgreSQL JDBC driver placed into `/kafka/libs` directory
+** https://docs.confluent.io/current/connect/connect-jdbc/docs/index.html[Kafka Connect JDBC Connector] placed into `/kafka/connect/kafka-connect-jdbc` directory
+* Pre-populated MySQL used in our link:docs/tutorial[tutorial]
+* Empty PostgreSQL
+
+There are certain impedances between Debezium MySQL Connector and Kafka Connect JDBC connector
+
+* event format
+* delete records/tombstone records handling
+* topic naming
+
+We are going to go through all those issues in detail.
+
+=== Event format
+Debezium emits events in a complex format that contains two record values: a value of the row record before the change and value of the row record after the change is applied.
+Further metadata are also included and the result is so called `Envelope` format.
+
+[source,json,indent=0]
+----
+{
+	"op": "u",
+	"source": {
+		...
+	},
+	"ts_ms" : "...",
+	"before" : {
+		"field1" : "oldvalue1",
+		"field2" : "oldvalue2"
+	},
+	"after" : {
+		"field1" : "newvalue1",
+		"field2" : "newvalue2"
+	}
+}
+----
+
+Kafka Connect JDBC connector (as most of the others) emits outgoing and expects incoming events in a simple format that is generally a new row record.
```

Comment:
*The* Kafka...

Reply:
Actually, as I mention in my comments, it's _Confluent's_ connector.